### PR TITLE
ENH: add fuzzy search

### LINKED ISSUE #52: fuzzy search
We need a fuzzy search capacity for both acq and an. @sbanerjee23

### PR DESCRIPTION
**Pull request by CJ-Wright**
Fuzzy wuzzy was a search,
Fuzzy wuzzy doesn't leave users in the lerch
Closes #52

### CODE REVIEW

**Comment by CJ-Wright** [comment]
@sbanerjee23 @sbillinge @tacaswell Ready for review!

**Comment by CJ-Wright** [comment]
The testing code starts [here](https://github.com/xpdAcq/xpdAn/pull/53/files#diff-81ea7915ef77f3db4fb842091d5677daR5).
We test that the code properly retrieves the correct headers even with slightly off search parameters.

**Comment by sbillinge** [comment]
Thanks Chris,

It is really the workflow that I am interested in testing....does it make sense for the user rather than does it pass tests (though we need those too....thanks).  If we can sit together tomorrow for a few minutes, preferrably with access to a database with something in it, we can give it test-drive.

**Comment by CJ-Wright** [comment]
Of course.

**Comment by CJ-Wright** [comment]
from @sbillinge need `pi_Last` in example for unknown name so we can evaluate the results.

File: xpdan/search.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+    -------
+    list:
+        A list
+
+    """
+    if isinstance(keys, list):
+        scores = []
+        for h in db():
+            scores.append(ndld(getFromDict(h['start'], keys), search_string))
+        scores = [ndld(getFromDict(h['start'], keys), search_string) for h in
```
Comment (tacaswell) [inline xpdan/search.py]: why do you compute this twice?

File: xpdan/search.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+                  db()]
+    else:
+        scores = [ndld(h['start'][keys], search_string) for h in db()]
+    zipped = zip(scores, db())
```
Comment (tacaswell) [inline xpdan/search.py]: there is a race condition here if an event gets added to the db between the first call and now.  It is better to grab the uid on the first pass.  In the next line you will pull everything into memory anyway so if you are going to do it this way, just make a list of header at the start.  

A better option is to keep a fixed size heap (see the `heapq` module) of the number you want to return, then do a single pass through the headers.  At each header use `heapq.heappushpop` https://docs.python.org/3.5/library/heapq.html#heapq.heappushpop to only keep the currntly seen `N` best.  As the last step, sort and return the relevant headers. 

I suggest using a heap element that looks like `(ndl(), header)`.

**Comment by tacaswell** [comment]
The 10k$ question is if it actually made it faster ;)

**Comment by tacaswell** [comment]
Hmm, do not quite understand what is going wrong here as 

```python
In [40]: (1, {',':5}) < (2, {">": 75})
Out[40]: True
```

while

```pytb
In [41]: {',':5} < {">": 75}
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-41-c7863bd416fb> in <module>()
----> 1 {',':5} < {">": 75}

TypeError: unorderable types: dict() < dict()

```

You might have to wrap the headers in an extra dict layer?  This might work better with the 'new' (vaporware) headers.  You can also just hold onto the uids and re-search them at the end.

**Comment by CJ-Wright** [comment]
In conversation with @tacaswell and for posterity:
The issue is that if the NL distances are the same then the comparison moves on to the headers which breaks.

File: xpdan/search.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+        if isinstance(v, dict):
+            yield from nested_dict_values(v)
+        else:
+            if isinstance(v, str):
+                yield v
+            else:
+                yield None
+
+
+def fuzzy_search(db, keys, search_string, size=100):
```
Comment (CJ-Wright) [inline xpdan/search.py]: Refactor this to take in kwargs as searches
```python
hdrs = fuzzy_search(db, piLast='foo')
```
This is much closer to databroker search syntax. Actually we might want to mirror it entirely with the exception of having a db arg.

File: xpdan/search.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+            The number of results to be returned, if 'all' all are returned.
+             Defaults to 100 results
+
+        Returns
+        -------
+        list:
+            A list
+
+        """
+    values = set([h['start'][key] for h in db()])
```
Comment (CJ-Wright) [inline xpdan/search.py]: heapify this as well

File: xpdan/search.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+    print: bool
+        If true prints the information
+
+    Returns
+    -------
+    list of dicts:
+        The list of beamtimes and their associated information
+    """
+    hdrs = db()
+    safs = set([h['start'][saf_key] for h in hdrs])
```
Comment (CJ-Wright) [inline xpdan/search.py]: It might be nice to give back the returns in reverse date order.

File: xpdan/search.py [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+    for h in db():
+        internal_scores = [1. - ndld(v, search_string) for v in
+                           nested_dict_values(h['start']) if v is not None]
+        heappushpop(heap, (max(internal_scores), h['start']['time'] * -1, h))
+    heap.sort()
+    heap.reverse()
+    return [g[-1] for g in heap if g[0] != -1]
+
+
+def beamtime_dates(db, keys=('facility', 'beamline', 'bt_safN'),
```
Comment (CJ-Wright) [inline xpdan/search.py]: Put the saf out first

**Comment by CJ-Wright** [comment]
@tacaswell and @danielballan should we subclass the databroker and attach the search functionality to it? @sbillinge raised a good point that it is a bit odd that `search` and `fuzzy_search` have different signatures (the `size` kwarg aside).

**Comment by danielballan** [comment]
That's up to you. Databroker is not in a stable enough state to make long-term promises about maintaining internal details, so just do what you think is best.

File: xpdan/search.py [hunk 7]
```diff
@@ -0,0 +1,358 @@
...
+        heapify(heap)
+        values = set([h['start'][key] for h in self()])
+        for v in values:
+            heappushpop(heap, (1. - ndld(v, search_string), v))
+        heap.sort()
+        heap.reverse()
+        return [g[-1] for g in heap if g[0] >= 0.]
+
+
+try:
```
Comment (tacaswell) [inline xpdan/search.py]: Don't do this, we have learned this is a bad pattern.  If you want to have an auto-configure module make a new module _just_ for the singleton.

File: xpdan/search.py [hunk 8]
```diff
@@ -0,0 +1,358 @@
...
+    from metadatastore.mds import MDSRO
+except (KeyError, ImportError) as exc:
+    warnings.warn("No default DataBroker object will be created because "
+                  "the necessary configuration was not found: %s" % exc)
+
+else:
+    DataBroker = FuzzyBroker(MDSRO(mds_config), FileStoreRO(fs_config))
+
+
+def get_from_dict(data_dict, map_list):
```
Comment (tacaswell) [inline xpdan/search.py]: consider making some of these private?

File: xpdan/glbl.py [hunk 9]
```diff
@@ -151,6 +151,9 @@ class Glbl:
     return Glbl
 
 
-env_code = os.environ['XPDAN_SETUP']
+try:
+    env_code = os.environ['XPDAN_SETUP']
```
Comment (CJ-Wright) [inline xpdan/glbl.py]: Do you mind if I deal with this in a separate PR?

**Comment by CJ-Wright** [comment]
@tacaswell @sbillinge I think this is ready for merge unless anyone has major objections.

**Pull request merged**